Support scaling HPA to/from zero pods for object/external metrics

DXist opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature


**What this PR does / why we need it**:
This PR targets worker deployments that use queues and scaling is based on object or external metric that depends on queue size. When workers are idle it is possible to scale corresponding deployment to zero replicas and save resources.

This technique is especially useful when workers request GPU resources and the amount of different idling worker types exceeds number of available GPUs.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #69687

**Special notes for your reviewer**:
The PR is based on changes made in https://github.com/kubernetes/kubernetes/pull/61423

1. Scale to/from zero changes are made
2. Applied changes from https://github.com/kubernetes/kubernetes/pull/61423
3. HPA continues to scale as long as there is at least one metric value available. 
There is no conservative scale down behaviour introduced in https://github.com/kubernetes/kubernetes/pull/61423
Scaling down works even if we have just one metric value.
4. HPA tolerance set through `--horizontal-pod-autoscaler-tolerance` flag is ignored when scaling up from zero pods.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
When HPAScaleToZero feature gate is enabled HPA supports scaling to zero pods based on object or external metrics. HPA remains active as long as at least one metric value available.

To downgrade the cluster to version that does not support scale-to-zero feature:
1. make sure there are no hpa objects with minReplicas=0. Here is a oneliner to update it to 1:
    $ kubectl get hpa --all-namespaces  --no-headers=true | awk  '{if($6==0) printf "kubectl patch hpa/%s --namespace=%s -p \"{\\\"spec\\\":{\\\"minReplicas\\\":1}}\"\n", $2, $1 }' | sh
2. disable HPAScaleToZero feature gate

```

DXist committed:
HPA scale-to-zero for custom object/external metrics

Add support for scaling to zero pods

minReplicas is allowed to be zero

condition is set once

Based on https://github.com/kubernetes/kubernetes/pull/61423

set original valid condition

add scale to/from zero and invalid metric tests

Scaling up from zero pods ignores tolerance

validate metrics when minReplicas is 0

Document HPA behaviour when minReplicas is 0

Documented minReplicas field in autoscaling APIs

DXist commented:
/assign @maciekpytel @thockin

spiffxp commented:
/ok-to-test

Rajat-0 commented:
@DXist  I am also working on this issue and I have discussed the below proposel in Sig weekly meeting, and as per discussion in meeting the Sig owner will discuss this with networking team, and based on that we can modify the PR and I'll be happy to coordinate with you on this.

https://docs.google.com/document/d/1p_Xlk8e5V32WOBVeRbJqbxhEKcJy_mDu9JgVp5ZNDxs/

DXist commented:
@rajat-0

> @DXist I am also working on this issue and I have discussed the below proposel in Sig weekly meeting, and as per discussion in meeting the Sig owner will discuss this with networking team, and based on that we can modify the PR and I'll be happy to coordinate with you on this.
> 
> https://docs.google.com/document/d/1p_Xlk8e5V32WOBVeRbJqbxhEKcJy_mDu9JgVp5ZNDxs/

This PR handles more general case and not specific to HTTP workload.

mwielgus reviewed (commented):
Couple of high-level items:

1. How was it tested?
2. Why didn't you handle: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L567
3. Did you consider explicitly validating that there is Object/External metric configured?
4. Maybe we should have a comment update in the API.

DXist commented:
> Couple of high-level items:
> 
> 1. How was it tested?

There are unittests that cover scaling up from zero pods and scaling down to zero for custom object and external metrics.

For end-to-end testing I

* built K8S binaries
* run them using https://github.com/kubernetes-sigs/kubeadm-dind-cluster
* set up RabbitMQ, Prometheus Operator, Prometheus adapter for custom metrics
* set up the service I want to scale and configured Prometheus, Prometheus adapter and HPA.
HPA used two Object metrics - worker utilization and queue ingress/egress ratio
* run scripts to generate test load - a single message or a series of messages in the rate of ~5X processing throughput of a single pod

Only one metric value (worker utilization) was available when there were no pods.
Metrics used 1m window to calculate averaged value so I delayed start of pod readiness probes via `initialDelaySeconds: 60`. Due to this delay the next scaling decision was based on a full metric window since previous rescale. The delay worked in the same way as [replica calculator code](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/replica_calculator.go#L381) for resource metrics.
> 2. Why didn't you handle: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L567

Github collapsed the diff for this module. I've added [extra condition](https://github.com/kubernetes/kubernetes/pull/74526/files#diff-ff4e8f1a0c24ee394a0ddf37d64b2960R587) to enable scaling when `minReplicas==0`
> 3. Did you consider explicitly validating that there is Object/External metric configured?

No, I haven't. It seems to be a good protection from HPA misconfiguration.

> 4. Maybe we should have a comment update in the API.

A comment update should be near [MinReplicas](https://github.com/kubernetes/kubernetes/blob/99f319567a5148f501e49da35c83478303eab38b/staging/src/k8s.io/api/autoscaling/v2beta2/types.go#L58)?

DXist commented:
Added metrics validation for `minReplicas=0` case and a comment to the API.

thockin commented:
API LGTM

@mwielgus ping me when LGTM and I will approve

mwielgus commented:
You may want to update api descriptions in 
https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/api/autoscaling as well.

mwielgus commented:
The PR seems OK. However, as we are late in 1.14 lifecycle and this is in fact an API change/new feature the PR has to wait until 1.15.

DXist commented:
@mwielgus , I've updated API descriptions for `autoscaling/v1`, `autoscaling/v2beta1` and `autoscaling/v2beta`.

DXist commented:
/retest

DXist committed:
Update generated files
update generated protobufs

swagger docs are regenerated

update openapi-spec

update generated openapi

DXist commented:
@mwielgus , I'm not sure how to fix this. Tried to run `make update` locally and got `working tree clean`

DXist commented:
/retest

DXist commented:
Rebased onto current master. API verification succeded.

DXist commented:
/retest

DXist commented:
@mwielgus , tests are passed.

Rajat-0 commented:
/lgtm

veonua commented:
please

DXist commented:
Rebased and squashed commits

mwielgus commented:
/lgtm

DXist commented:
/retest

josephburnett commented:
/lgtm

Nice change!

mwielgus commented:
/lgtm
/approve

DXist commented:
@thockin , your approve is required

mwielgus commented:
@thockin Can you approve the api change, please :) ?

thockin reviewed (commented):
I want to discuss with API reviewers.  Strictly speaking this is a breaking change, but I am inclined to allow this class of changes.

/hold

thockin commented:
Hi all.

API reviewers had a lively discussion about this, since it is strictly a breaking change.  The short form is that this is a low-impact "break" and everyone is OK with it in principle, but to roll out a breaking change there's a nascent staging process we ought to follow.

https://github.com/kubernetes/kubernetes/pull/69415#issuecomment-461094828

thoughts?

DXist commented:
@thockin , thank you for your time for review.

Are there any options for a backward data migration for cluster state during downgrade process? 
In this case the migration could iterate over HPA objects and unset `minReplicas` if it equals to zero to maintain compatible state.

If no then in release 1.15 we can introduce a feature gate to allow `minReplicas` to be zero for new HPA objects. And make the feature as generally available in 1.16.

lavalamp commented:
@DXist No-- no data manipulation is done in that scenario. If you want that behavior, you have to make a new version. Then you can write whatever you want in a conversion function.

DXist commented:
> @DXist No-- no data manipulation is done in that scenario. If you want that behavior, you have to make a new version. Then you can write whatever you want in a conversion function.

@lavalamp if I understand correctly we have to handle the case with k8s rollback, haven't we?
The case assumes following steps:
- install K8S cluster with version X+1 (1.15 or 1.16)
- create an HPA object with `minReplicas=0` and configured custom metrics
- rollback the cluster to version X (1.14 or 1.15)
- deal with `minReplicas=0` in HPA internal representation stored in etcd

I haven't tested this scenario yet.

mwielgus commented:
@DXist - Did you manage to test the scenario?

DXist commented:
> @DXist - Did you manage to test the scenario?

@mwielgus , no I haven't.

If I understood @lavalamp correctly API object validation happens only on write path. Rollback to 1.14 should break update operations for scale-to-zero HPA objects until `minReplicas` is set to a valid value.

Is it a major issue? Can we just document this behaviour during rollback in release notes?

mwielgus commented:
@lavalamp @thockin What should we do with this PR?

mikkeloscar commented:
> Is it a major issue? Can we just document this behaviour during rollback in release notes?

:+1:

mwielgus commented:
@lavalamp @thockin  Ping :)

lavalamp commented:
> Rollback to 1.14 should break update operations for scale-to-zero HPA objects until minReplicas is set to a valid value.

Including writes to status, and some delete paths (since they do updates)... This is a fairly severe form of breakage which is very confusing to users when it happens. I am not sure what the probability of this happening is.

mwielgus commented:
@lavalamp @thockin I understand that there is some risk of breakage, but we need a decision. Is this PR a GO or NO-GO?

DXist commented:
We discussed in Slack introduction of a feature gate to allow scale-to-zero. I don't see any value in this gate since we don't introduce new API features.

Instead we can document the need to update HPA minReplicas to nonzero value in case cluster operator decides to downgrade the cluster.

```
Rinat Shigapov [8:25 PM]

I guess there should be not many users who will try scaling to zero pod and then do a rollback.
Do you recommend to introduce a feature gate to allow validation code to store ‘0’ value for minReplicas? And make the new feature as GA in 1.16?

lavalamp [8:40 PM]
what happens when the feature gate is turned off?

Rinat Shigapov [8:43 PM]
Autoscaling validation code will not allow to set 0 to minReplicas.
At the moment it allows it if there is a custom or object metric configured for a HPA object.

lavalamp [8:52 PM]
Yeah that sounds safer than rolling back

Rinat Shigapov [10:39 AM]
Just to make sure we understand each other: when I said about the rollback I meant any decision to downgrade the cluster not necessary related to this feature.

Rinat Shigapov [11:06 AM]
Let’s imagine an operator who installed K8S 1.16 with GA scale-to-zero support.

The operator created an HPA object with minReplicas=0 and configured a custom metric.

Then by some reason the operator decided to downgrade the cluster to 1.15, which should allow scale-to-zero only if the flag is turned on.

If the feature flag is turned off (which is default) the situation with invalid minReplicas would be the same as current.

So I don’t see any value of introducing this flag.

lavalamp [6:27 PM]
To roll back safely, you can turn off the flag, touch all HPA objects, and then roll back

Rinat Shigapov [7:49 PM]
Or just touch HPA objects - update minReplicas to nonzero value. And then roll back.

Rinat Shigapov [3:11 PM]
The minimal required enhancement of the PR is to document the need to update minReplicas (if it’s zero) before K8S downgrade.

mikkeloscar commented:
> Instead we can document the need to update HPA minReplicas to nonzero value in case cluster operator decides to downgrade the cluster.

:+1:

lavalamp commented:
> Instead we can document the need to update HPA minReplicas to nonzero value in case cluster operator decides to downgrade the cluster.

I would prefer not to do this, it's too manual.

Instead I recommend:
* have a feature gate, which, when off, always forces the value to be at least 1.
* document that to roll back, one needs to first turn the feature gate off, then touch all HPA objects (you can e.g. provide a bash + kubectl example script which does this), then start the prior object
* alternatively, on rollback, just accept that some HPA objects will not be modifyable (including to set status) until this field is fixed up.

lavalamp commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -103,14 +103,28 @@ func ValidateHorizontalPodAutoscalerStatusUpdate(newAutoscaler, oldAutoscaler *a
+		}
+	}
+
+	if minReplicas != nil && *minReplicas == 0 {
```
Should be if it's nil or zero?

DXist replied:
We need at least one object or external metric configured if minReplicas is set to zero. Not nil check is required to avoid nil pointer dereference.

lavalamp replied:
Officially it defaults to 1, but the defaulting code should have already run by the time this function is invoked. So, it's not clear what nil means at this point, I was assuming 0...

DXist replied:
@lavalamp minReplicas is a pointer to int32.

I've done the same nil protective check as [here](https://github.com/kubernetes/kubernetes/pull/74526/files/1e9ec7b634181f4f9220d0cc10c9381b3e410986#diff-4364e3f754b12e39f60eea33105f6c7eR47)

It looks robust and doesn't require to know about external defaulting code.

liggitt replied:
this check looks fine. practically, it will not be nil since defaulting runs before we get here, but the nil check makes it clearly safe

timmycarr commented:
Hi! Friendly reminder from your release team: We are starting the code freeze for 1.15 tomorrow EOD. Seeing some good progress here! Just checking in to see if this is still planned for the 1.15 cycle?

mwielgus commented:
@timmycarr I really hope we will manage to squeeze it in at the very last moment. 

@lavalamp - Just to be on the safe side, are your current comment the only concerns that you have?

thockin commented:
I see no feature gates in here.  I can't tell if any planning has been done to mitigate the breakingness of the change.  If so, can you explain?

Even what Daniel suggests (https://github.com/kubernetes/kubernetes/pull/74526#issuecomment-496625514) seems too loose to me.  How many of those changes could we tolerate in a release?  Certainly not more than 1 or 2.

I can't approve a breaking change, regardless of the relative unlikeliness of it triggering.  We have to hold the bar.

Please ping me ASAP if you think this is wrong.

liggitt commented:
Agreed. https://github.com/kubernetes/kubernetes/pull/69415#issuecomment-461094828 is a good overview of how to do this (quoted and expanded here for reference):

In 1.15:

1. handle the expanded values, and tolerate persisted objects with the expanded values:
  1a: update validation to validate the expanded values
  1b: update consumers of the field to handle the expanded values
  1c: allow an update of an object already using the expanded values to continue using them
2. prevent new use of the expanded values (unless they opt into an alpha feature gate):
  2a: prevent creation of a new object using the expanded values
  2b: prevent an update of an object not yet using the expanded values from starting to use them

In 1.16, the restrictions on create and update (2a,2b) can be removed.

Making people who want to use this in 1.15 explicitly opt into an alpha feature protects unaware users from using a feature in a way that breaks HA clusters involving 1.14 and 1.15 servers, or breaks them on rollback.

DXist commented:
/retest

soggiest commented:
/milestone v1.16

mikkeloscar commented:
> /milestone v1.16

:'(

DXist commented:
@thockin @lavalamp @mwielgus I've rebased onto master and added `HPAScaleToZero` feature gate.

thockin reviewed (commented):
I think this is sane now.  @liggitt want another look?

/approve

alok87 commented on pkg/controller/podautoscaler/replica_calculator.go:
```diff
@@ -241,20 +241,30 @@ func (c *ReplicaCalculator) GetObjectMetricReplicas(currentReplicas int32, targe
+		replicaCount = int32(math.Ceil(usageRatio * float64(readyPodCount)))
+	} else {
+		// Scale to zero or n pods depending on usageRatio
+		replicaCount = int32(math.Ceil(usageRatio))
```
This logic is not working in our case. We are scaling our deployment from min to max based on SQS queue metric. Suppose we want to scale the deployment `sqsworker` from `min=1` to `max=20` pods. Scaling should happen when the queue length is greater than 100.

With this logic when we scale down to 0. There is no way to come to 1 untill the queue length becomes > 100.

```
ApproximateNumberOfMessages = 0
    usageRatio = 0/100 = 0
    currentReplicas = 1
    readyPodCount = 1
        desired = int32(math.Ceil(usageRatio * float64(readyPodCount)))
                = int32(0*1) = 0

    ApproximateNumberOfMessages = 2
    usageRatio = 2/100 = 0.02
    currentReplicas = 0
    readyPodCount = 0
        desired = int32(math.Ceil(usageRatio * float64(readyPodCount)))
                = int32(ceil(0.02*0)) = 0
```

alok87 commented on pkg/controller/podautoscaler/replica_calculator.go:
```diff
@@ -241,20 +241,30 @@ func (c *ReplicaCalculator) GetObjectMetricReplicas(currentReplicas int32, targe
+		replicaCount = int32(math.Ceil(usageRatio * float64(readyPodCount)))
+	} else {
+		// Scale to zero or n pods depending on usageRatio
+		replicaCount = int32(math.Ceil(usageRatio))
```
We have a use case in which we need to scale worker deployments from min=0 to min=100 (for example) based on the SQS queue length. The worker deployments works on the SQS queue.

Do you see this logic working with AWS SQS? SQS does not have a single metric to give the usageRatio because of the long polling of the queue by the consumers, the `ApproximateMessage` may be showing 0 all the time.

DXist replied:
@alok87 , personally I use this code with RabbitMQ and Prometheus. I have an heuristic of worker utilization which depends on the number of messages in the queue and the number of workers.

felippe-mendonca replied:
@DXist, sorry but wasn't clear to me which Prometheus exporter for RabbitMQ you did use on tests, was [prometheus_rabbitmq_exporter](https://github.com/deadtrickster/prometheus_rabbitmq_exporter) or [rabbitmq_exporter](https://github.com/kbudde/rabbitmq_exporter). Moreover, what was the metric used on that? In a rollback scenario which you doesn't have scaling HPA to/from zero feature available, it'll still work? Does it scale to one instead of zero?

DXist replied:
@fellippe-mendonca, I use the second exporter which is configured in [rabbitmq-ha helm chart](https://github.com/helm/charts/blob/master/stable/rabbitmq-ha/values.yaml#L465)

I have 2 metrics - worker utilisation and ingress/egress message rate ratio. [Here](http://ix.io/1Okl) is a Helm template of PrometheusRule object I've configured.

If you want to rollback the cluster to version without scale-to-zero support you have to update HPA objects explicitly before the rollback. See release notes to this PR.

felippe-mendonca replied:
Thanks for the information @DXist, I'll try this approach with an application I'm running with RabbitMQ that doesn't scale well with CPU. I was wondering to use average response time, but I think that ingress/egress rate ratio might be better! Hope this new functionality come to k8s ASAP and further for GKE.

DXist commented:
I've rebased onto current master.

liggitt commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -44,8 +47,20 @@ var ValidateHorizontalPodAutoscalerName = apivalidation.ValidateReplicationContr
+	var MinReplicasLowerBound int32
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.HPAScaleToZero) {
+		MinReplicasLowerBound = 1
```
validateHorizontalPodAutoscalerSpec is called from ValidateHorizontalPodAutoscalerUpdate.

Once we enable this feature by default (e.g. 1.17), an API server will allow minReplicas of 0. In a skewed HA apiserver scenario, the prior version of the apiserver with the feature disabled (e.g. 1.16) would reject updates to the object the 1.17 allowed to be persisted. If a 1.16 server encounters a persisted object that already has minReplicas set to 0, it must allow updates of that object with minReplicas=0.

One way to do this:
* pass in `minReplicasLowerBound` as a parameter to validateHorizontalPodAutoscalerSpec.
* when called from `ValidateHorizontalPodAutoscaler` (on creation), compute `minReplicasLowerBound` the way you do here (based only on the feature gate)
* when called from `ValidateHorizontalPodAutoscalerUpdate` (on update), allow a lower bound of 0 if the feature gate is enabled OR if the existing HPA has minReplicas=0.

liggitt replied:
This also needs unit tests exercising all of these scenarios:
* feature enabled, ValidateHorizontalPodAutoscaler() allows minReplicas=0
* feature disabled, ValidateHorizontalPodAutoscaler() forbids minReplicas=0
* feature enabled, old HPA has minReplicas=1, ValidateHorizontalPodAutoscalerUpdate() allows minReplicas=0
* feature enabled, old HPA has minReplicas=0, ValidateHorizontalPodAutoscalerUpdate() allows minReplicas=0
* feature disabled, old HPA has minReplicas=1, ValidateHorizontalPodAutoscalerUpdate() forbids minReplicas=0
* feature disabled, old HPA has minReplicas=0, ValidateHorizontalPodAutoscalerUpdate() allows minReplicas=0

DXist replied:
@liggitt I've updated validation logic and moved `minReplicas` tests into separate [test functions](https://github.com/kubernetes/kubernetes/pull/74526/files#diff-f9431b410b931831536d0896b59f1c44R1083) to isolate feature gate toggling.

liggitt commented on pkg/apis/autoscaling/validation/validation_test.go:
```diff
@@ -96,6 +99,10 @@ func TestValidateHorizontalPodAutoscaler(t *testing.T) {
 	}
+
+	// Enable HPAScaleToZero feature gate.
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.HPAScaleToZero, true)()
```
rather than enabling unconditionally for all tests, we need tests exercising behavior with the feature enabled and disabled

DXist replied:
I moved the gate switching to separate test functions.

liggitt commented on staging/src/k8s.io/api/autoscaling/v1/types.go:
```diff
@@ -38,7 +38,9 @@ type HorizontalPodAutoscalerSpec struct {
 	ScaleTargetRef CrossVersionObjectReference `json:"scaleTargetRef" protobuf:"bytes,1,opt,name=scaleTargetRef"`
-	// lower limit for the number of pods that can be set by the autoscaler, default 1.
+	// MinReplicas is the lower limit for the number of replicas to which the autoscaler can scale down.
+	// It defaults to 1 pod. MinReplicas is allowed to be 0 if at least one Object or External metric is configured.
```
nit: use `minReplicas` to match the serialized name API consumers will see.

include a mention that setting minReplicas=0 requires the alpha feature gate HPAScaleToZero to be enabled.

DXist replied:
Updated comments in all API versions.

liggitt commented:
> I think this is sane now. @liggitt want another look?

this is close, but needs a couple tweaks to handle skew cases properly. left a description at https://github.com/kubernetes/kubernetes/pull/74526/files#r302972197

DXist commented:
/test pull-kubernetes-bazel-test

DXist commented:
@liggitt I've updated the PR.

liggitt commented on pkg/apis/autoscaling/validation/validation_test.go:
```diff
@@ -1002,3 +1006,158 @@ func TestValidateHorizontalPodAutoscaler(t *testing.T) {
+		if errs := ValidateHorizontalPodAutoscalerUpdate(&nonZeroCase, &zeroCase); len(errs) != 0 {
+			t.Errorf("expected success: %v", errs)
+		}
+	}
```
needs one final case where both old and new are zero, and update validation should pass

DXist replied:
@liggitt I've added extra check to cover this case.

liggitt commented:
validation logic looks correct. one test case missing: `feature disabled, old HPA has minReplicas=0, ValidateHorizontalPodAutoscalerUpdate() allows minReplicas=0`

liggitt commented:
/hold cancel
validation looks good, will lgtm once test case is added

liggitt commented:
/lgtm
/retest

DXist commented:
/retest

DXist commented:
@liggitt , @mwielgus What is the next step to make the feature GA?

numbsafari commented:
Just stopping by after six months to see what, if anything, can be done to move this otherwise perfectly reasonable and working feature from Alpha to GA.

This would be great to have on hosted K8S environments like GKE or AKS or EKS or... basically anywhere.

numbsafari commented:
Was just reading the release notes for 1.19,
and noticed the section on [avoiding permanent beta](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.19.md#avoiding-permanent-beta), it would be great if there was some clarity on how to foster this work through the process.

jeffreybrowning commented:
> @liggitt , @mwielgus What is the next step to make the feature GA?

Would love an answer to this. It's been half a year or so since it was asked. 

Very useful feature for shutting down workers processing jobs from queues when the queues have no tasks.

lavalamp commented:
Sounds like a good topic for the SIG Autoscaling meeting :)

jeffreybrowning commented:
@lavalamp was there results from the meeting?

lavalamp commented:
Apologies, that was super unclear of me-- I realize now it reads like I was going to take it to the SIG meeting, but I was actually only trying to suggest that it was the logical next thing to do. I don't personally have time to contribute to this other than the occasional review.

jeffreybrowning commented:
@lavalamp no worries. Is there someone more central to the planning we can poke here to move things along? I think this should be a pretty innocent bump to GA, if not Beta. I'll go ahead and tag others who were in this ticket for Alpha release.

@liggitt @mwielgus

johanneswuerbach commented:
We are also using this feature heavily and I would be happy to contribute time driving this towards GA. 

While I contributed some changes to k8s, I've never been part of an API graduation so I would need to some pointers how to do this if help is needed.

liggitt commented:
> I think this should be a pretty innocent bump to GA. I'll go ahead and tag others who were in this ticket for Alpha release.
> 
> @liggitt @mwielgus

API changes need an associated doc describing the change and graduation requirements, called a [KEP](https://github.com/kubernetes/enhancements/tree/master/keps) (Kubernetes Enhancement Proposal). The key things we'd be looking for are evidence of sufficient test coverage, any performance implications of the change, and upgrade/downgrade/skew compatibility implications.

It looks like https://github.com/kubernetes/kubernetes/issues/69687#issuecomment-467082733 might have been the original proposal associated with the alpha change, but before graduating to beta and being enabled by default, that should be ported to a KEP and make sure the test/performance/upgrade questions are answered. A template is available at https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template and would be placed in https://github.com/kubernetes/enhancements/tree/master/keps/sig-autoscaling

jeffreybrowning commented:
@johanneswuerbach @DXist is this something one of you can lead? It would be my first contribution, and is probably not a great thing to take a bite out of first.

If not, I can attempt the first pass at the proposal.

DXist commented:
@jeffreybrowning go ahead. I changed project and work in another context.

johanneswuerbach commented:
I planned to get something drafted on Friday, but I also never worked on a KEP before. If you have more time feel free to pick it @jeffreybrowning :-)

jeffreybrowning commented:
I am embroiled in an update to another package right now -- it would likely be a sizeable delay until I get to this.

johanneswuerbach commented:
I started the official propose here https://github.com/kubernetes/enhancements/issues/2021. Looking forward to any of your comments or feedback on the KEP itself here https://github.com/kubernetes/enhancements/pull/2022.
